sdks/java/core/src/main/java/org/apache/beam/sdk/util/HistogramData.java

```diff
@@ -227,6 +276,133 @@ public interface BucketType extends Serializable {
     double getAccumulatedBucketSize(int endIndex);
   }
 
+  @AutoValue
+  public abstract static class ExponentialBuckets implements BucketType {
+
+    public abstract double getBase();
+
+    public abstract int getScale();
+
+    /**
+     * Set to 2**scale which is equivalent to 1/log_2(base). Precomputed to use in {@code
+     * getBucketIndexPositiveScale}
+     */
+    public abstract double getInvLog2GrowthFactor();
+
+    @Override
+    public abstract int getNumBuckets();
+
+    /* Precomputed since this value is used everytime a datapoint is recorded. */
+    @Override
+    public abstract double getRangeTo();
+
+    public static ExponentialBuckets of(int scale, int numBuckets) {
+      if (scale < -3) {
+        throw new IllegalArgumentException(
+            String.format("Scale should be greater than -3: %d", scale));
+      }
+
+      if (scale > 3) {
+        throw new IllegalArgumentException(String.format("Scale should be less than 3: %d", scale));
+      }
+      if (numBuckets <= 0) {
+        throw new IllegalArgumentException(
+            String.format("numBuckets should be greater than zero: %d", numBuckets));
+      }
+
+      double invLog2GrowthFactor = Math.pow(2, scale);
+      double base = Math.pow(2, Math.pow(2, -scale));
+      int clippedNumBuckets = ExponentialBuckets.computeNumberOfBuckets(scale, numBuckets);
+      double rangeTo = Math.pow(base, clippedNumBuckets);
+      return new AutoValue_HistogramData_ExponentialBuckets(
+          base, scale, invLog2GrowthFactor, clippedNumBuckets, rangeTo);
+    }
+
+    /**
+     * numBuckets is clipped so that the largest bucket's lower bound is not greater than 2^31-1
+     * (uint32 max).
+     */
+    private static int computeNumberOfBuckets(int scale, int inputNumBuckets) {
+      if (scale == 0) {
+        // When base=2 then the bucket at index 31 contains [2^31, 2^32).
+        return Math.min(32, inputNumBuckets);
```

Comment:
same as comment above, maybe rename and declare as `private static final int MAX_INPUT_NUM_BUCKETS = 32` inside `ExponentialBuckets`